File: android/app/src/main/java/com/audiobookshelf/app/data/TTSBook.kt
package com.audiobookshelf.app.data

import com.fasterxml.jackson.annotation.JsonIgnoreProperties

/**
 * Extracted ebook text for the native read aloud (TTS) player.
 * Produced by the ebook readers in the WebView (ttsExtractBook hooks)
 * and cached on disk by TTSBookCache.
 * See docs/native-tts-player-design.md
 */
@JsonIgnoreProperties(ignoreUnknown = true)
data class TTSBook(
  var libraryItemId: String,
  var serverAddress: String?,
  var title: String,
  var author: String?,
  var language: String,
  var rate: Float,
  // TTS engine package name and voice name; null/empty = engine keeps its current values
  var ttsEngine: String? = null,
  var voice: String? = null,
  var ebookFormat: String,
  var chapters: MutableList<TTSChapter>,
  var totalChars: Int,
  /**
   * Page skips from the media session (notification, lock screen, Android
   * Auto): pages per skip and the reader's estimate of characters on one
   * displayed page (0 = unknown, the engine falls back to a default).
   */
  var pageStep: Int = DEFAULT_PAGE_STEP,
  var pageChars: Int = 0
) {
  constructor() : this("", null, "", null, "en-US", 1f, null, null, "", mutableListOf(), 0)

  companion object {
    const val DEFAULT_PAGE_STEP = 3
    const val DEFAULT_PAGE_CHARS = 1500
  }

  /** Total characters before the given position, for progress and time estimates */
  fun charsBefore(chapterIndex: Int, paragraphIndex: Int): Int {
    var chars = 0
    chapters.forEachIndexed { ci, chapter ->
      chapter.paragraphs.forEachIndexed { pi, paragraph ->
        if (ci < chapterIndex || (ci == chapterIndex && pi < paragraphIndex)) {
          chars += paragraph.chars
        }
      }
    }
    return chars
  }

  /** Position of the paragraph with the given saved ebookLocation, or null when not found */
  fun positionForLocation(location: String?): Pair<Int, Int>? {
    if (location.isNullOrEmpty()) return null
    chapters.forEachIndexed { ci, chapter ->
      chapter.paragraphs.forEachIndexed { pi, paragraph ->
        if (paragraph.location == location) return Pair(ci, pi)
      }
    }
    return null
  }

  /** Position of the paragraph containing the given ebookProgress ratio (by character count) */
  fun positionForProgress(progressRatio: Double): Pair<Int, Int> {
    val targetChars = (progressRatio.coerceIn(0.0, 1.0) * totalChars).toInt()
    var chars = 0
    chapters.forEachIndexed { ci, chapter ->
      chapter.paragraphs.forEachIndexed { pi, paragraph ->
        chars += paragraph.chars
        if (chars > targetChars) return Pair(ci, pi)
      }
    }
    val lastChapterIndex = maxOf(0, chapters.size - 1)
    return Pair(lastChapterIndex, maxOf(0, (chapters.lastOrNull()?.paragraphs?.size ?: 1) - 1))
  }
}

@JsonIgnoreProperties(ignoreUnknown = true)
data class TTSChapter(
  var title: String?,
  var startLocation: String?,
  var paragraphs: MutableList<TTSParagraph>,
  /**
   * Chapter start as an epub cfi, when the extraction could build one. Saved as
   * the reading position for paragraphs without their own cfi - unlike the
   * spine href in [startLocation] every ebook reader can resume from it.
   */
  var startCfi: String? = null
) {
  constructor() : this(null, null, mutableListOf())
}

@JsonIgnoreProperties(ignoreUnknown = true)
data class TTSParagraph(
  var text: String,
  var location: String?,
  var chars: Int
) {
  constructor() : this("", null, 0)
}

/** Lightweight summary stored next to the full book for fast listing (Android Auto browse) */
@JsonIgnoreProperties(ignoreUnknown = true)
data class TTSBookSummary(
  var libraryItemId: String,
  var serverAddress: String?,
  var title: String,
  var author: String?,
  var language: String,
  var ebookFormat: String,
  var totalChars: Int,
  var lastAccessed: Long
) {
  constructor() : this("", null, "", null, "en-US", "", 0, 0L)
}

File: ios/App/Shared/models/TTSBook.swift
//
//  TTSBook.swift
//  Audiobookshelf
//
//  Extracted ebook text for the native read aloud (TTS) player - the same
//  payload the Android engine speaks (android/.../data/TTSBook.kt). Produced
//  by the ebook readers in the WebView (ttsExtractBook hooks) and cached on
//  disk by TTSBookCache. See docs/native-tts-player-design.md
//

import Foundation

struct TTSBook: Codable {
    static let defaultPageStep = 3
    static let defaultPageChars = 1500

    var libraryItemId: String
    var serverAddress: String?
    var title: String
    var author: String?
    var language: String
    var rate: Float
    /// Engine package name (Android only) and voice identifier; nil/empty = keep the current values
    var ttsEngine: String?
    var voice: String?
    var ebookFormat: String
    var chapters: [TTSChapter]
    var totalChars: Int
    /// Page skips from the lock screen / remote commands: pages per skip and the
    /// reader's estimate of characters on one displayed page (0 = unknown, the
    /// engine falls back to a default)
    var pageStep: Int
    var pageChars: Int

    private enum CodingKeys: String, CodingKey {
        case libraryItemId, serverAddress, title, author, language, rate, ttsEngine, voice, ebookFormat, chapters, totalChars, pageStep, pageChars
    }

    init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        libraryItemId = try values.decodeIfPresent(String.self, forKey: .libraryItemId) ?? ""
        serverAddress = try values.decodeIfPresent(String.self, forKey: .serverAddress)
        title = try values.decodeIfPresent(String.self, forKey: .title) ?? ""
        author = try values.decodeIfPresent(String.self, forKey: .author)
        language = try values.decodeIfPresent(String.self, forKey: .language) ?? "en-US"
        rate = try values.decodeIfPresent(Float.self, forKey: .rate) ?? 1
        ttsEngine = try values.decodeIfPresent(String.self, forKey: .ttsEngine)
        voice = try values.decodeIfPresent(String.self, forKey: .voice)
        ebookFormat = try values.decodeIfPresent(String.self, forKey: .ebookFormat) ?? ""
        chapters = try values.decodeIfPresent([TTSChapter].self, forKey: .chapters) ?? []
        totalChars = try values.decodeIfPresent(Int.self, forKey: .totalChars) ?? 0
        pageStep = try values.decodeIfPresent(Int.self, forKey: .pageStep) ?? TTSBook.defaultPageStep
        pageChars = try values.decodeIfPresent(Int.self, forKey: .pageChars) ?? 0
    }

    func chapter(at chapterIndex: Int) -> TTSChapter? {
        guard chapterIndex >= 0, chapterIndex < chapters.count else { return nil }
        return chapters[chapterIndex]
    }

    func paragraph(chapterIndex: Int, paragraphIndex: Int) -> TTSParagraph? {
        guard let chapter = chapter(at: chapterIndex), paragraphIndex >= 0, paragraphIndex < chapter.paragraphs.count else { return nil }
        return chapter.paragraphs[paragraphIndex]
    }

    /// Total characters before the given position, for progress and time estimates
    func charsBefore(chapterIndex: Int, paragraphIndex: Int) -> Int {
        var chars = 0
        for (ci, chapter) in chapters.enumerated() {
            for (pi, paragraph) in chapter.paragraphs.enumerated() {
                if ci < chapterIndex || (ci == chapterIndex && pi < paragraphIndex) {
                    chars += paragraph.chars
                }
            }
        }
        return chars
    }

    /// Position of the paragraph with the given saved ebookLocation, or nil when not found
    func position(forLocation location: String?) -> (chapterIndex: Int, paragraphIndex: Int)? {
        guard let location = location, !location.isEmpty else { return nil }
        for (ci, chapter) in chapters.enumerated() {
            if let pi = chapter.paragraphs.firstIndex(where: { $0.location == location }) {
                return (ci, pi)
            }
        }
        return nil
    }

    /// Position of the paragraph containing the given ebookProgress ratio (by character count)
    func position(forProgress progressRatio: Double) -> (chapterIndex: Int, paragraphIndex: Int) {
        let targetChars = Int(min(max(progressRatio, 0), 1) * Double(totalChars))
        var chars = 0
        for (ci, chapter) in chapters.enumerated() {
            for (pi, paragraph) in chapter.paragraphs.enumerated() {
                chars += paragraph.chars
                if chars > targetChars {
                    return (ci, pi)
                }
            }
        }
        return lastPosition
    }

    /// Last paragraph of the book ((0, 0) for an empty book)
    var lastPosition: (chapterIndex: Int, paragraphIndex: Int) {
        let lastChapterIndex = max(0, chapters.count - 1)
        return (lastChapterIndex, max(0, (chapters.last?.paragraphs.count ?? 1) - 1))
    }
}

struct TTSChapter: Codable {
    var title: String?
    var startLocation: String?
    /// Chapter start as an epub cfi, when the extraction could build one. Saved as
    /// the reading position for paragraphs without their own cfi - unlike the
    /// spine href in startLocation every ebook reader can resume from it.
    var startCfi: String?
    var paragraphs: [TTSParagraph]

    private enum CodingKeys: String, CodingKey {
        case title, startLocation, startCfi, paragraphs
    }

    init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        title = try values.decodeIfPresent(String.self, forKey: .title)
        startLocation = try values.decodeIfPresent(String.self, forKey: .startLocation)
        startCfi = try values.decodeIfPresent(String.self, forKey: .startCfi)
        paragraphs = try values.decodeIfPresent([TTSParagraph].self, forKey: .paragraphs) ?? []
    }
}

struct TTSParagraph: Codable {
    var text: String
    var location: String?
    var chars: Int

    private enum CodingKeys: String, CodingKey {
        case text, location, chars
    }

    init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        text = try values.decodeIfPresent(String.self, forKey: .text) ?? ""
        location = try values.decodeIfPresent(String.self, forKey: .location)
        // The readers count UTF-16 units (String.length in JS) - keep the same unit
        chars = try values.decodeIfPresent(Int.self, forKey: .chars) ?? (text as NSString).length
    }
}

/// Lightweight summary stored next to the full book for fast listing
struct TTSBookSummary: Codable {
    var libraryItemId: String
    var serverAddress: String?
    var title: String
    var author: String?
    var language: String
    var ebookFormat: String
    var totalChars: Int
    /// Unix time in ms of the last save/load, for the LRU eviction
    var lastAccessed: Double
}
